Using pixelCopy to screen capture screenshot remotely in an Assurance session

cacheung opened the pull request:
Using pixelCopy to screen capture screenshot remotely Assurance session for Android device with API 26+, Android 8.0+, so it fixes a crash issue while taking screenshot for hardware rendering view.

For the older devices (API 21-25, Android 5.0-7.1), we use Canvas method to support software rendering views.
We do not use Drawing Cache method for software rendering because it is deprecated in API 28.

Add erroring messaging for older version (Below API 21, Android 4.4 and below).

Add MAX_SCREENSHOT_SIZE to prevent OutOfMemoryError for large screen.

Update unit tests.

<!--- Provide a general summary of your changes in the Title above -->

## Description

<!--- Describe your changes in detail -->

## Related Issue

<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->

## Motivation and Context

<!--- Why is this change required? What problem does it solve? -->

## How Has This Been Tested?

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Screenshots (if appropriate):

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

cacheung committed:
Fix Screenshot Crash for hardware, using pixelCopy

Fix Screenshot Crash hardware, using pixelCopy

cacheung committed:
fix spotless format

fix spotless format

prudrabhat commented on code/assurance/src/main/java/com/adobe/marketing/mobile/assurance/internal/AssurancePluginScreenshot.java:
```diff
@@ -89,26 +81,94 @@ private void getCurrentScreenShot(final CaptureScreenShotListener captureScreenS
+
+        // Check if device supports PixelCopy API (Android 8.0 and above)
+        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) {
+            Log.error(
+                    Assurance.LOG_TAG,
+                    LOG_TAG,
+                    "Screenshot not supported on Android versions below 8.0 (API 26)");
+            if (parentSession != null) {
+                parentSession.logLocalUI(
+                        UILogColorVisibility.LOW,
+                        "Screenshot not supported on Android versions below 8.0 (API 26)");
+                sendErrorEvent("Screenshot not supported on Android versions below 8.0 (API 26)");
+            }
```
This will regress existing screenshot compatibility with `API 21` to `API 26` that we currently support (albeit not the hardware bitmaps). 
Can we fallback to default screenshot capturing for `21 <= API <= 25` and then fail with an error event if it turns out to be a hardware bitmap?

```
if (Build.VERSION.SDK_INT >= 21 && Build.VERSION.SDK_INT <= 25 ) {
   try {
     // success for software bitmap
   } catch (e) {
      // fail here for hardware bitmap
     }
 
} else {
      
      // capture screenshot with PixelCopy

}


```

cacheung committed:
Fallback to use Canvas based screenshot for Android 21 - 25.

Fallback to use Canvas based screenshot for Android 21 - 25.
Use pixelCopy for Android 26 and above.
Provide error message for Android version below 21.

cacheung committed:
format fix

format fix

prudrabhat commented on code/assurance/src/main/java/com/adobe/marketing/mobile/assurance/internal/AssurancePluginScreenshot.java:
```diff
@@ -89,26 +85,215 @@ private void getCurrentScreenShot(final CaptureScreenShotListener captureScreenS
+
+    @RequiresApi(api = Build.VERSION_CODES.O)
+    private void captureScreenshotWithPixelCopy(final Activity currentActivity) {
+        currentActivity.runOnUiThread(
```
Would you be able to check if it is indeed required to call `PixelCopy.request()` on main thread? Since we pass the main thread handler, I feel it should be OK to call it from the current thread of invocation.

cacheung replied:
I did some research, the PixelCopy.request() doesn't need main thread.  The view property access (getWidth(), getHeight(), getLocationOnScreen()) before the call requires UI thread per Android guidelines.  Since PixelCopy.request() is already async/non-blocking, there's no performance benefit to moving it off main thread, only potential crash risk from unsafe view access.

prudrabhat commented on code/assurance/src/main/java/com/adobe/marketing/mobile/assurance/internal/AssurancePluginScreenshot.java:
```diff
@@ -89,26 +85,215 @@ private void getCurrentScreenShot(final CaptureScreenShotListener captureScreenS
+                parentSession.logLocalUI(
+                        UILogColorVisibility.LOW,
+                        "Screenshot not supported on Android versions below 5.0 (API 21)");
+                sendErrorEvent("Screenshot not supported on Android versions below 5.0 (API 21)");
```
This else block is not required because we require apps to be min SDK 21 to use our extension.

prudrabhat commented on code/assurance/src/main/java/com/adobe/marketing/mobile/assurance/internal/AssurancePluginScreenshot.java:
```diff
@@ -28,9 +35,10 @@ class AssurancePluginScreenshot implements AssurancePlugin {
     private static final String PAYLOAD_MIMETYPE = "mimeType";
     private static final String PAYLOAD_ERROR = "error";
-    private AssuranceSession parentSession = null;
+    private static final int MAX_SCREENSHOT_SIZE =
```
Let's add unit to the constant measure so that it is easier to know if this bytes or pixels.

prudrabhat commented on code/assurance/src/main/java/com/adobe/marketing/mobile/assurance/internal/AssurancePluginScreenshot.java:
```diff
@@ -89,26 +85,215 @@ private void getCurrentScreenShot(final CaptureScreenShotListener captureScreenS
+                            int width = view.getWidth();
+                            int height = view.getHeight();
+
+                            // Limit bitmap size to prevent OutOfMemoryError
```
Let's move this scaling calculation to a private method since it appears that it is  being used in 2 places.

cacheung committed:
Incorporate review comments

Incorporate review comments

cacheung committed:
Spotless applied

Spotless applied
